gramps/gen/lib/json_utils.py

```diff
@@ -0,0 +1,201 @@
+            value = self[key]
+            if isinstance(value, dict):
+                return DataDict(value)
+            elif isinstance(value, list):
+                return DataList(value)
+            else:
+                return value
+        else:
+            self["_object"] = data_to_object(self)
+            return getattr(self["_object"], key)
+
+
+class DataList(list):
+    """
+    A wrapper around a data list.
+    """
+
+    def __getitem__(self, position):
+        value = super().__getitem__(position)
+        if isinstance(value, dict):
+            return DataDict(value)
+        elif isinstance(value, list):
+            return DataList(value)
+        else:
+            return value
+
+
+def convert_state_to_object(obj_dict):
+    _class = obj_dict.pop("_class")
+    cls = lib.__dict__[_class]
+    obj = cls.__new__(cls)
+    obj.set_object_state(obj_dict)
+    return obj
+
+
+# Just call the method directly:
+# def convert_object_to_state(obj):
+#     return obj.get_object_state()
+
+
+def string_to_data(string: str | bytes) -> DataDict:
+    """
+    Convert a JSON string into its data representation.
+    """
+    return DataDict(orjson.loads(string))
+
+
+def string_to_dict(string: str | bytes) -> dict:
+    """
+    Convert a JSON string into its dict representation.
+    """
+    return orjson.loads(string)
+
+
+def dict_to_string(dict: dict) -> str | bytes:
+    """
+    Convert a dict into its JSON string representation.
+    """
+    return orjson.dumps(dict)
+
+
+def object_to_dict(obj):
+    """
+    Convert any Gramps lib object into its DataDict representation.
+    """
+    if isinstance(obj, (float, int, str, NoneType)):
+        return obj
+
+    elif isinstance(obj, (list, tuple)):
+        return [object_to_dict(item) for item in obj]
+
+    state = obj.get_object_state()
+    return {k: object_to_dict(v) for k, v in state.items()}
+
+
+def object_to_data(obj):
+    """
+    Convert any Gramps lib object or other value into
+    its dict representation or value, respectively.
+    """
+    if isinstance(obj, (float, int, str, NoneType)):
+        return obj
+
+    elif isinstance(obj, (list, tuple)):
+        return [object_to_data(item) for item in obj]
+
+    state = obj.get_object_state()
+    return {k: object_to_data(v) for k, v in state.items()}
+
+
+def data_to_object(data):
+    """
+    Convert any object dict representation to a Gramps lib object
+    or other value.
+    """
+    if isinstance(data, dict):
+        if "_object" in data:
+            return data["_object"]
+
+        data = {k: data_to_object(v) for k, v in data.items()}
+        return convert_state_to_object(data)
+
+    elif isinstance(data, (list, tuple)):
+        return [data_to_object(item) for item in data]
+
+    return data
+
+
+def object_to_string(obj: object) -> str | bytes:
+    """
+    Convert any Gramps object into a JSON string/bytes.
+    """
+    return orjson.dumps(object_to_dict(obj))
```

Comment:
I'm thinking out loud here: would having a custom `JSONEncoder` class allow us to skip the `object_to_dict` call?
See the `cls` parameter https://docs.python.org/3/library/json.html#basic-usage

Reply:
Not a crazy idea. Which do you think is faster?

1. json.dumps(obj, MyCustomEncoder) - Python converts all dict items to strings
2. orjson.dumps(object_to_dict(obj)) - Python converts objs to dict, orjson does the rest

I am always surprised when I guess, but I bet #2 is faster. Let's find out! Place your bets.

Hey, it looks like [orjson.dumps()](https://github.com/ijl/orjson?tab=readme-ov-file#serialize) accepts `default`. I didn't notice that before. There may be a third option here.

Reply:
Nice. I missed the support for `default`.

Reply:
> Which do you think is faster?

well I felt a customencoder should be quicker if it avoids creating the full object hierarchy as a dict, only to then traverse the dict to convert to JSON. There should be fewer memory allocations this way. But the python allocator is quick. I await your test results with interest.....

Reply:
I didn't use the class, but the default, which is what we started with (results in seconds, on a 40k person table):

1. `json.dumps(default=convert_object_to_state)`: 6.143402099609375
2. `orjson.dumps(object_to_dict(obj))`: 5.892540454864502
3. `orjson.dumps(person, default=convert_object_to_state)` 4.168360948562622

So, option 3 wins! Another 40% faster! I'll make that change tomorrow.

Reply:
I'll try the class API too, for completeness.

Reply:
4. `json.dumps(person, cls=MyEncoder)`: 6.621660470962524

I'll go with 3.

Here is the class for inspection:

```python
class MyEncoder(json.JSONEncoder):
    def default(self, obj):
        if obj.__class__.__module__.startswith("gramps"):
            obj = obj.get_object_state()
            return self.default(obj)
        elif isinstance(obj, (list, tuple)):
            return str([self.default(item) for item in obj])
        else:
            return obj
```